Add prompt for telemetry consent

fvictorio opened the pull request:
This prompt shows in two scenarios:

- When setting up a project (but only for the sample project, not for the empty config)
- The first time a command is executed, except the help command (either explicit, or because the user ran `npx hardhat` without args)

The prompts is only shown if the config file doesn't exist. After responding, this is saved in that file as a JSON like `{ "consent": true }`.

I used a separate file, instead of a general `config.json` with entries for each setting, just because I couldn't find a proper way of atomically modifying an on-disk JSON, so it seems more robust to have separate files. But maybe this is being too cautious.

alcuadrado commented on packages/hardhat-core/src/internal/util/global-dir.ts:
```diff
@@ -10,12 +11,23 @@ async function generatePaths() {
 
+function generatePathsSync() {
+  const envPaths: typeof envPathsT = require("env-paths");
+  return envPaths("hardhat");
```
When did we change these to `hardhat`? We should be really careful about this.

fvictorio replied:
I think this was part of the rename? Anyway, there's a `TODO-HH` below that should remind us to take the migration into consideration.

alcuadrado replied:
oops, i missed it

alcuadrado commented on packages/hardhat-core/src/internal/util/global-dir.ts:
```diff
@@ -75,3 +87,29 @@ export async function getCompilersDir() {
+ *
+ * Returns undefined if it can't be determined.
+ */
+export function hasConsentedTelemetry(): boolean | undefined {
```
Wouldn't the whole pr be simpler if this were async?

fvictorio replied:
Yes, but that would make the error reporter async, and since it's used in the constructor of `provider/node.ts`, it would mean having to find a way around that.

alcuadrado replied:
Oh, it's because of the error reporter! I thought it was because of the project creation module, which is already async, so that confused me.

alcuadrado commented on packages/hardhat-core/src/internal/cli/project-creation.ts:
```diff
@@ -367,6 +377,23 @@ async function confirmRecommendedDepsInstallation(): Promise<boolean> {
   return responses.shouldInstallPlugin === true;
 }
 
+export async function confirmTelemetryConsent(): Promise<boolean> {
```
IMO this function doesn't belong into this file.

alcuadrado reviewed (commented):
I left a few comments, but apart form them, this looks good.

alcuadrado commented:
Some tests fail now because there's no telemetry consent in the CI. Not sure what should be done there.

fvictorio commented:
@alcuadrado I fixed the tests and added one extra one to cover the case where the user hasn't given telemetry consent. Do you think that's enough, or is there something more that we should be testing here?

alcuadrado commented on packages/hardhat-core/src/internal/cli/analytics.ts:
```diff
@@ -150,8 +148,8 @@ export class Analytics {
       // Custom dimension 1: Project Id
-      // This is computed as the keccak256 hash of the project's absolute path.
-      cd1: this._projectId,
+      // TODO-HH: Can we remove this?
```
I don't think we can. Debugging these things is super cumbersome also.

fvictorio replied:
Removed the comment and used "hardhat-project" instead of "project" as the first dimension.

alcuadrado commented on packages/hardhat-core/src/internal/cli/project-creation.ts:
```diff
@@ -361,6 +371,23 @@ async function confirmRecommendedDepsInstallation(): Promise<boolean> {
+      name: "telemetryConsent",
+      type: "confirm",
+      initial: true,
+      message: "Can Hardhat send anonymous data to improve the product?",
```
```suggestion
      message: "Help us improve Hardhat with anonymous crash reports & basic usage data?",
```

alcuadrado commented:
Merged this with a linter error. Will fix it rn.
